### ISSUE TITLE
max_open_connections minimum value causes multiple failures

### ISSUE THREAD

**Issue by archen** [hashicorp/boundary#1374]
**Describe the bug**
Following [the configuration docs](https://www.boundaryproject.io/docs/configuration/controller#max_open_connections), I set my `max_open_connections` in the database stanza to 2. This leads to multiple failures in the boundary controller across feature sets; and inconsistent behavior [between the CLI and official Terraform provider](https://github.com/hashicorp/terraform-provider-boundary/issues/126).

**To Reproduce**
1. Configure boundary.hcl with `max_open_connections = 2` in the database stanza
2. Try to provision almost any of the terraform examples across hashi learn-boundary-* repos
3. `boundary_role` resources that modify `grant_strings` will fail
4. Try to modify targets in the web ui as an administrator, failures
5. Try to `boundary connect` to targets, failures

**Expected behavior**
I expected the `max_open_connections = 2` as a minimum value to be sufficient to operate a boundary deployment.

**Actual behavior**
This is the output and terraform on a `terraform apply` _after_ having provisioned a functional controller
https://gist.github.com/archen/4f62f96642374509c6ab4cc4073590e8

Boundary status and cli output from connect commands
https://gist.github.com/archen/2381080191e0da5b74dc1e198284f8f0

I was going to capture the process and logs for modifying the target in UI, but forcing the two prior failure conditions caused the server to go unresponsive (assuming a resource deadlock somewhere).

**Additional context**
All of my nodes are arm64. The issue occurred across multiple combinations of versions:
- boundary (0.3.0 and 0.4.0)
- terraform provider (1.0.1 and 1.0.2)
- postgresql (11 and 13)

Removing the `max_open_connections` line from config resolves all of the issues.

**Comment by malnick** [comment]
Hi @archen - thank you for raising this. We're in the process of reproducing this, but we think it's related to how Terraform is provisioning resources. By default, Terraform uses a `parallelism` of 10: https://www.terraform.io/docs/cli/commands/apply.html#parallelism-n

With the DB connection limit set to 2, Boundary is unable to connect to PG while its API might have 10 connections for CRUD operations coming from Terraform. We'll report back with more steadfast testing results soon, thank you!

**Comment by archen** [comment]
Thanks @malnick 

I suspect there's more going on with the connection pools. Terraform was the most consistent (and earliest) manifestation of the problem that I observed, but even after getting the roles/targets/etc configured with the CLI, I was having issues with Boundary functioning until I removed the `max_open_connections` setting:
- Unable to modify resources in the UI
- Backend orchestration failing between the controller and worker (which I assume also caused my inability to connect to targets with the client)

The behavior persisted through reboots, attaching to fresh databases, etc. I'm confident that my execution was consistent as the only methods I used were Ansible and Terraform for the deployment/bootstrapping. 

Removing `max_open_connections` with no other change to the deployment method or configuration appears to be the only reliable way to avoid the failures.
